neothesia/src/scene/playing_scene/midi_player.rs

```diff
@@ -198,69 +252,202 @@ pub enum MidiEventSource {
 
         if active {
-            self.user_pressed_recently
-                .push_back(UserPress { timestamp, note_id });
-            self.required_notes.remove(&note_id);
+            // Check if note_id has reached required_notes, then remove it now
+            if let Some(index) = self
+                .required_notes
+                .iter()
+                .position(|item| item.note_id == note_id)
+            {
+                if timestamp
+                    .duration_since(self.required_notes[index].timestamp)
+                    .as_millis()
+                    > 160
+                {
+                    //160 to forgive touching the bottom
+
+                    self.user_stats.notes_missed += 1;
+                } else {
+                    self.user_stats.notes_hit += 1;
+                }
+                self.required_notes.remove(index);
+
+                self.user_notes.push_back(UserPress {
+                    timestamp: timestamp,
+                    note_id,
+                    occurrence: *occurrence,
+                    time_key_up: timestamp,
+                });
+            } else {
+                // Haven't reached required_notes yet, place a possible later validation in 'user_pressed_recently' / file_press_key()
+                if let Some(item) = self
+                    .user_pressed_recently
+                    .iter_mut()
+                    .find(|item| item.note_id == note_id)
+                {
+                    // already exists, update timestamp
+                    item.timestamp = timestamp;
+                    self.user_stats.wrong_notes += 1;
+                } else {
+                    // Not found, push a new UserPress
+                    self.user_pressed_recently.push_back(UserPress {
+                        timestamp,
+                        note_id,
+                        occurrence: *occurrence,
+                        time_key_up: timestamp,
+                    });
+                }
+            }
+        } else {
+            // Update user_notes log time_key_up
+            if let Some(item) = self
+                .user_notes
+                .iter_mut()
+                .rev()
+                .find(|item| item.note_id == note_id && item.occurrence == *occurrence)
+            {
+                item.time_key_up = Instant::now();
+            }
         }
     }
 
     fn file_press_key(&mut self, note_id: u8, active: bool) {
+        let occurrence = self.occurrence.entry(note_id).or_insert(0);
+        let timestamp = Instant::now();
         if active {
-            if let Some((id, _)) = self
+            *occurrence += 1;
+
+            // Check if note got pressed earlier 500ms (user_pressed_recently)
+            if let Some((_id, item)) = self
                 .user_pressed_recently
                 .iter()
                 .enumerate()
                 .find(|(_, item)| item.note_id == note_id)
```

Comment:
Why enumerate if you never use the index?
```suggestion
            if let Some(item) = self
                .user_pressed_recently
                .iter()
                .find(|item| item.note_id == note_id)
```